producer.go

```diff
@@ -247,11 +194,40 @@ func (p *segmentProducer) Close() error {
 	return nil
 }
 
-func (p *segmentProducer) getID(m kafka.Message) string {
-	for _, h := range m.Headers {
-		if h.Key == kafkaMessageIDHeader {
-			return string(h.Value)
-		}
-	}
-	return ""
+// batchSizeAdjustingBalancer is a balancer that adjusts the batch size of the
+// writer based on the number of messages that we want to write.
+// Before calling Writer.WriteMessages you can call SetRecordCount on the
+// balancer to let it know how many messages will be written. First time the
+// Balance method is called it will calculate the appropriate batch size based
+// on the number of partitions and adjust the setting in the writer.
+// The actual balancing logic after that is delegated to kafka.RoundRobin.
+//
+// When the record count is divisible by the number of partitions, this balancer
+// will ensure that all batches are flushed instantly. If it's not divisible and
+// the record count is greater than the number of partitions, some batches will
+// be flushed only after the batch delay is reached.
+//
+// This is a workaround and best approximation for achieving synchronous writes,
+// because the segment client doesn't allow you to write without batching.
+// For more info see https://github.com/segmentio/kafka-go/issues/852.
+type batchSizeAdjustingBalancer struct {
```

Comment:
Nitpick: It would be good to mention that the Segment Writer's batch size is per-partition.